### PR TITLE
Import identifiers from DwCA

### PR DESCRIPTION
**Pull request by charvolant**
Also, cleaner resolution of vernacular names, taking synonyms into account.

This extension is needed by the lists tool to be able to follow lists where the "names" are actually taxon identifiers.

Also, a fix for annoying issue https://github.com/AtlasOfLivingAustralia/ala-name-matching/issues/140

### CODE REVIEW

File: ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java [hunk 1]
```diff
@@ -219,20 +219,20 @@ public NameKey analyse(@Nullable NomenclaturalClassifier code, String scientific
...
         }
+        try {
+            name = this.nameParser.parse(scientificName, (rankType == null || rankType == RankType.UNRANKED) ? null : rankType.getCbRank());
+            if (name != null) {
+                nameType = name.getType();
+                if (rankType == null && name.getRank() != null)
+                    rankType = RankType.getForCBRank(name.getRank());
+            }
+        } catch (UnparsableException ex) {
+            // Oh well, worth a try
```
Comment (nickdos) [inline ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java:230]: Just checking we might want to `log.error` this line for debugging prod issues ??

Comment (charvolant) [reply ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java:230]: This part of the pre-prod merge. If a name is unparsable, it is just treated as being "as-is" without extra information from parsing. Having said that, an log.info might be appropriate.

However, hang on before approving. I want to fix the "Genus B" problem, as well.

File: ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java [hunk 2]
```diff
@@ -299,7 +299,18 @@ public NameKey analyse(@Nullable NomenclaturalClassifier code, String scientific
         scientificName = scientificName.trim().toUpperCase();
 
 
-        return new NameKey(this, code, scientificName, scientificNameAuthorship, rankType, nameType, flags);
+        NameKey key = new NameKey(this, code, scientificName, scientificNameAuthorship, rankType, nameType, flags);
+        if (name == null)
```
Comment (nickdos) [inline ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java:303]: Reminder: un-parentheses `if/else` syntax is discouraged according to our coding std ([oracle standard](https://www.oracle.com/java/technologies/javase/codeconventions-statements.html#449))

File: ala-name-matching-search/src/test/java/au/org/ala/names/search/AutocompleteTest.java [hunk 3]
```diff
@@ -150,13 +151,13 @@ public void testAutocomplete12() throws Exception {
...
-        assertNotNull(synonyms);
-        assertTrue(synonyms.size() > 0);
-        Map synonym = synonyms.get(0);
-        assertEquals("Yellow Rush Lily", synonym.get("commonname"));
+        Optional<Map> syn = results.stream().filter(r -> "Sisyrinchium micranthum".equals(r.get("name"))).findFirst();
+        assertTrue(syn.isPresent());
+        //List<Map> synonyms = (List<Map>) syn.get().get("synonymMatch");
+        //assertNotNull(synonyms);
+        //assertTrue(synonyms.size() > 0);
+        //Map synonym = synonyms.get(0);
```
Comment (nickdos) [inline ala-name-matching-search/src/test/java/au/org/ala/names/search/AutocompleteTest.java:159]: Dave prefers deleting code as git history can be used to get it back

File: ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java (hunk 2 above)
Comment (charvolant) [reply ala-name-matching-builder/src/main/java/au/org/ala/names/index/ALANameAnalyser.java]: Fixed with change to logic

File: ala-name-matching-search/src/test/java/au/org/ala/names/search/AutocompleteTest.java (hunk 3 above)
Comment (charvolant) [reply ala-name-matching-search/src/test/java/au/org/ala/names/search/AutocompleteTest.java]: Removed.

**Review by nickdos — APPROVED** [review]

**Pull request merged**